lib/obj.js

```diff
@@ -1,365 +1,479 @@
-                    var ui = getOffset(u, uvs, 2);
-                    var ux = uvs[ui + 0];
-                    var uy = uvs[ui + 1];
-                    // Flip y so 0.0 is the bottom of the image
-                    uy = 1.0 - uy;
-                    vertexArray.push(ux, uy);
-                } else {
-                    // Some objects in the model may not have uvs, fill with 0's for consistency
-                    vertexArray.push(0.0, 0.0);
-                }
-            }
+        // UVs
+        if (defined(u)) {
+            var ui = getOffset(u, uvs, 2);
+            var ux = uvs.get(ui + 0);
+            var uy = uvs.get(ui + 1);
+            mesh.uvs.push(ux);
+            mesh.uvs.push(uy);
         }
+    }
 
-        function addVertex(v, p, u, n) {
-            var index = vertexCache[v];
-            if (!defined(index)) {
-                index = vertexCount++;
-                vertexCache[v] = index;
-                createVertex(p, u, n);
+    function addVertex(v, p, u, n) {
+        var index = vertexCache[v];
+        if (!defined(index)) {
+            index = vertexCount++;
+            vertexCache[v] = index;
+            createVertex(p, u, n);
+
+            // Prevent the vertex cache from growing too large. As a result of clearing the cache there
+            // may be some duplicate vertices.
+            vertexCacheCount++;
+            if (vertexCacheCount > vertexCacheLimit) {
+                vertexCacheCount = 0;
+                vertexCache = {};
             }
+        }
+        return index;
+    }
 
-            return index;
+    function addFace(v1, p1, u1, n1, v2, p2, u2, n2, v3, p3, u3, n3, v4, p4, u4, n4) {
+        var index1 = addVertex(v1, p1, u1, n1);
+        var index2 = addVertex(v2, p2, u2, n2);
+        var index3 = addVertex(v3, p3, u3, n3);
+
+        primitive.indices.push(index1);
+        primitive.indices.push(index2);
+        primitive.indices.push(index3);
+
+        // Triangulate if the face is a quad
+        if (defined(v4)) {
+            var index4 = addVertex(v4, p4, u4, n4);
+            primitive.indices.push(index1);
+            primitive.indices.push(index3);
+            primitive.indices.push(index4);
         }
+    }
 
-        function addFace(v1, p1, u1, n1, v2, p2, u2, n2, v3, p3, u3, n3, v4, p4, u4, n4) {
-            var index1 = addVertex(v1, p1, u1, n1);
-            var index2 = addVertex(v2, p2, u2, n2);
-            var index3 = addVertex(v3, p3, u3, n3);
-
-            currentIndexArray.push(index1);
-            currentIndexArray.push(index2);
-            currentIndexArray.push(index3);
-
-            // Triangulate if the face is a quad
-            if (defined(v4)) {
-                var index4 = addVertex(v4, p4, u4, n4);
-                currentIndexArray.push(index1);
-                currentIndexArray.push(index3);
-                currentIndexArray.push(index4);
-            }
+    function parseLine(line) {
+        line = line.trim();
```

Comment:
See my similar regex comments in the mtl parsing code.  This is all really inefficient and not GC friendly.  We can handle it in a separate PR if you want.